pkg/pub/publication.go

```diff
@@ -11,40 +16,108 @@ type Publication struct {
 	Fetcher  fetcher.Fetcher   // The underlying fetcher used to read publication resources.
 	// TODO servicesBuilder
 	// TODO positionsFactory
-	// TODO services []Service
-	_manifest manifest.Manifest
+	services []service.Service
 }
 
 // Returns whether this publication conforms to the given Readium Web Publication Profile.
 func (p Publication) ConformsTo(profile manifest.Profile) bool {
 	return p.Manifest.ConformsTo(profile)
 }
 
+// Returns the RWPM JSON representation for this [Publication]'s manifest, as a string.
+func (p Publication) JSONManifest() (string, error) {
+	bin, err := json.Marshal(p.Manifest)
+	if err != nil {
+		return "", err
+	}
+	return string(bin), nil
+}
+
+// The URL where this publication is served, computed from the [Link] with `self` relation.
+func (p Publication) BaseURL() *string {
+	lnk := p.Manifest.Links.FirstWithRel("self")
+	if lnk == nil {
+		return nil
+	}
+	dir := path.Dir(lnk.Href)
+	return &dir
+}
+
+// Returns the first existing link matching the given [path].
+func (p Publication) Find(path string) *manifest.Link {
+	link := p.Manifest.Links.FirstWithHref(path)
+	if link == nil {
+		link = p.Manifest.ReadingOrder.FirstWithHref(path)
+		if link == nil {
+			link = p.Manifest.Resources.FirstWithHref(path)
+			if link == nil {
+				return nil
+			}
+		}
+	}
+
+	link.Href = "/" + link.Href
+	return link
+}
+
+// Returns the resource targeted by the given non-templated [link].
+func (p Publication) Get(link manifest.Link) fetcher.Resource {
+	for _, service := range p.services {
+		if l, ok := service.Get(link); ok {
+			return l
+		}
+	}
+	return p.Fetcher.Get(link)
+}
+
+// Free up resources associated with the publication
 func (p Publication) Close() {
 	p.Fetcher.Close()
+	for _, service := range p.services {
+		service.Close()
+	}
 }
 
-func New() *Publication {
-	return &Publication{}
+func New(m manifest.Manifest, f fetcher.Fetcher, b *service.ServicesBuilder) *Publication {
+	if b == nil {
+		b = service.NewBuilder(nil, nil, nil, nil, nil)
+	}
+	services := b.Build(service.NewContext(m, f)) // Build the services
+	var newManifest manifest.Manifest
+	copier.Copy(&newManifest, &m) // Make a deep copy of the manifest
```

Comment:
Why do we need to do a deep copy if the `Manifest` has value semantics? Are there nested reference properties?

Reply:
I'm replicating what was done here: https://github.com/readium/kotlin-toolkit/blob/540c80fc1c57f04d09db45764f103724177aedf9/readium/shared/src/main/java/org/readium/r2/shared/publication/Publication.kt#L88 which was to deep copy the manifest and add the service links

Reply:
Kotlin doesn't have struct/value objects unfortunately, everything works by reference. So we had to make an actual copy of the manifest.

But if I'm not mistaken we don't need it with Go, even if the struct contains arrays. I did a quick test: https://go.dev/play/p/GfOWdK7sWMK